File: packages/ui/Sources/UI/Badge/Badge.swift
import SwiftUI

public enum BadgeTone {
  case success
  case neutral
  case warning
  case danger

  var foreground: Color {
    switch self {
    case .success: .green
    case .neutral: .secondary
    case .warning: .orange
    case .danger: .red
    }
  }

  var background: Color {
    foreground.opacity(0.15)
  }
}

/// A small labeled pill, e.g. a "Registered" status tag. Generic — the caller supplies the text/tone.
public struct Badge: View {
  private let title: String
  private let tone: BadgeTone

  public init(_ title: String, tone: BadgeTone = .neutral) {
    self.title = title
    self.tone = tone
  }

  public var body: some View {
    Text(title)
      .font(.caption.weight(.semibold))
      .padding(.horizontal, 10)
      .padding(.vertical, 5)
      .background(tone.background)
      .foregroundColor(tone.foreground)
      .clipShape(Capsule())
  }
}

#Preview {
  VStack(spacing: 8) {
    Badge("Registered", tone: .success)
    Badge("Pending", tone: .warning)
    Badge("Neutral")
  }
  .padding()
}

File: packages/ui/Java/com/variance/nearby/ui/Badge.kt
package com.variance.nearby.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp

enum class BadgeTone {
    SUCCESS,
    NEUTRAL,
    WARNING,
    DANGER,
}

/** A small labeled pill, e.g. a "Registered" status tag. Generic — the caller supplies text + tone. */
@Composable
fun Badge(
    text: String,
    tone: BadgeTone = BadgeTone.NEUTRAL,
    modifier: Modifier = Modifier,
) {
    val foreground =
        when (tone) {
            BadgeTone.SUCCESS -> Color(0xFF1B873F)
            BadgeTone.NEUTRAL -> MaterialTheme.colorScheme.onSurfaceVariant
            BadgeTone.WARNING -> Color(0xFFB26A00)
            BadgeTone.DANGER -> MaterialTheme.colorScheme.error
        }
    Text(
        text = text,
        style = MaterialTheme.typography.bodySmall,
        fontWeight = FontWeight.Bold,
        color = foreground,
        modifier =
            modifier
                .clip(RoundedCornerShape(8.dp))
                .background(foreground.copy(alpha = 0.15f))
                .padding(horizontal = 10.dp, vertical = 5.dp),
    )
}
